### PR TITLE
Add support for BERT for relevance transfer

### PR DESCRIPTION
**Pull request by achyudh**
Commit history is garbled. Feel free to squash the changes.

### CODE REVIEW

**Comment by Ashutosh-Adhikari** [comment]
Can we please maintain a separate branch for this until EMNLP?

File: common/evaluators/relevance_transfer_evaluator.py [hunk 1]
```diff
@@ -4,73 +4,124 @@
...
 class RelevanceTransferEvaluator(Evaluator):
 
-    def __init__(self, dataset_cls, model, embedding, data_loader, batch_size, device, keep_results=False):
-        super().__init__(dataset_cls, model, embedding, data_loader, batch_size, device, keep_results)
-        self.ignore_lengths = False
+    def __init__(self, model, config, **kwargs):
+        super().__init__(kwargs['dataset'], model, kwargs['embedding'], kwargs['data_loader'],
+                         batch_size=config['batch_size'], device=config['device'])
+
+        if config['model'] in {'BERT-Base', 'BERT-Large'}:
```
Comment (daemon) [inline common/evaluators/relevance_transfer_evaluator.py:24]: Why not use `bert-large-uncased` or `bert-large` for `config['model']`?

File: common/trainers/relevance_transfer_trainer.py [hunk 2]
```diff
@@ -1,133 +1,154 @@
...
-                    self.early_stop = True
-                    tqdm.write("Early Stopping. Epoch: {}, Best Dev F1: {}".format(epoch, self.best_dev_ap))
-                    break
+        os.makedirs(os.path.join(self.model_outfile, self.config['dataset'].NAME), exist_ok=True)
+
+        if self.config['model'] in {'BERT-Base', 'BERT-Large'}:
+            train_features = convert_examples_to_features(
+                self.train_examples, self.config['max_seq_length'], self.tokenizer)
+
+            all_input_ids = torch.tensor([f.input_ids for f in train_features], dtype=torch.long)
```
Comment (daemon) [inline common/trainers/relevance_transfer_trainer.py:121]: `torch.(lowercase t)ensor` should already default to longs for `int` Python types.

File: datasets/bert_processors/robust45_processor.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    def __init__(self, input_ids, input_mask, segment_ids, label_id, guid):
+        super().__init__(input_ids, input_mask, segment_ids, label_id)
+        self.guid = guid
+
+
+class Robust45Processor(BertProcessor):
+    NAME = 'Robust45'
+    NUM_CLASSES = 2
+    TOPICS = ['307', '310', '321', '325', '330', '336', '341', '344', '345', '347', '350', '353', '354', '355', '356',
```
Comment (daemon) [inline datasets/bert_processors/robust45_processor.py:17]: This is redundant?

**Comment by achyudh** [comment]
Can someone please look at this? It's been open for a while now.

**Review by daemon — APPROVED** [review]

File: common/evaluators/relevance_transfer_evaluator.py (hunk 1 above)
Comment (achyudh) [reply common/evaluators/relevance_transfer_evaluator.py:24]: I just wanted to be consistent with the model names (KimCNN, XML-CNN etc.). I replace `BERT-Large` to `bert-large-uncased` in the driver method.

File: common/trainers/relevance_transfer_trainer.py (hunk 2 above)
Comment (achyudh) [reply common/trainers/relevance_transfer_trainer.py:121]: Maybe `train_features` contains Python float variables? I got an error downstream when calculating the loss and I had to explicitly specify `torch.long` here to resolve it.

File: datasets/bert_processors/robust45_processor.py (hunk 3 above)
Comment (achyudh) [reply datasets/bert_processors/robust45_processor.py:17]: Yeah it is. I can use the values from the Robust45 class, but then I would have to create a dict that maps the dataset classes with the corresponding processors. I'll add it to the things to do in the future.
